diff --git a/subsys/bluetooth/controller/ll_sw/ctrl.h b/subsys/bluetooth/controller/ll_sw/ctrl.h
index ecec3ccc3..88cdba5fc 100644
--- a/subsys/bluetooth/controller/ll_sw/ctrl.h
+++ b/subsys/bluetooth/controller/ll_sw/ctrl.h
@@ -128,7 +128,7 @@
 #if defined(CONFIG_BT_CTLR_COMPANY_ID)
 #define RADIO_BLE_COMPANY_ID            CONFIG_BT_CTLR_COMPANY_ID
 #else
-#define RADIO_BLE_COMPANY_ID            0xFFFF
+#define RADIO_BLE_COMPANY_ID (0x0059) // Nordic Semiconductor ASA
 #endif
 #if defined(CONFIG_BT_CTLR_SUBVERSION_NUMBER)
 #define RADIO_BLE_SUB_VERSION_NUMBER \
diff --git a/subsys/Kconfig b/subsys/Kconfig
index 4966a0b81..546713963 100644
--- a/subsys/Kconfig
+++ b/subsys/Kconfig
@@ -24,3 +24,5 @@ source "subsys/net/Kconfig"
 source "subsys/shell/Kconfig"
 
 source "subsys/usb/Kconfig"
+
+source "subsys/btstack/Kconfig"
diff --git a/subsys/Makefile b/subsys/Makefile
index a9a2aa153..94da28f17 100644
--- a/subsys/Makefile
+++ b/subsys/Makefile
@@ -1,6 +1,7 @@
 obj-$(CONFIG_FILE_SYSTEM) += fs/
 obj-$(CONFIG_USB) += usb/
 obj-$(CONFIG_BT) += bluetooth/
+obj-$(CONFIG_BTSTACK) += btstack/
 obj-$(CONFIG_NET_BUF) += net/
 obj-$(CONFIG_CONSOLE_SHELL) += shell/
 obj-$(CONFIG_CONSOLE_PULL) += console/
